### DISCUSSION TITLE
Upgrading from 1.15.2 to 1.15.3 gives error "Fatal Error: composer.lock was created for PHP version 7.3 or higher but the current PHP version is 7.2.2."

### DISCUSSION THREAD

**Discussion by RayCulp** [kimai/kimai#2822] [On-Premise: Installation & Updates]
**To Reproduce**
1. Execute the following commands via SSH:
`git fetch --tags`
`git checkout 1.15.3`
`composer install --no-dev --optimize-autoloader`
3. The following error is displayed:
![image](https://user-images.githubusercontent.com/7621330/136112941-347acabf-79d3-4a1b-b99a-593346cce90a.png)

**Additional context**
I'm on shared hosting. I opened a ticket with the provider a few months ago about an issue that is probably related, but never followed up because at the time it was just a warning, not an actual error:
 
![image](https://user-images.githubusercontent.com/7621330/136114411-cfe46a56-30c8-4030-88f9-cf08c04db169.png)

The problem is that I'm not explicitly calling any version of PHP at all. It seems to be implicitly called when I execute the command `composer install --no-dev --optimize-autoloader`

Any ideas how to solve this?

**Comment by RayCulp** [comment]
Also, possibly related: When I try to do a clean install of 1.15.1, 1.15.2 or 1.15.3 according to the instructions, I get this error message when executing `composer install --no-dev --optimize-autoloader`

![image](https://user-images.githubusercontent.com/7621330/136276256-f0498d00-5cf5-40ba-a282-14e0c617755e.png)

**Comment by kevinpapst** [comment]
You need to use the correct PHP version, which was answered by your support.
Read the docs to find out how to do it in general: 
https://www.kimai.org/documentation/installation.html#shared-hosting

**Comment by RayCulp** [comment]
> which was answered by your support

I'm afraid I don't see how they answered anything.

If I were calling PHP explicitly, for example something like this

`php composer install --no-dev --optimize-autoloader`

then I could use the correct PHP version like this

`php7.3 composer install --no-dev --optimize-autoloader`

That actually works perfectly for commands like  `php7.3 occ maintenance:mode --on` (for Nextcloud)

But in this case

1. I'm not calling PHP explicitly, so I can't preface the command with `php7.3`
2. I certainly do not have sufficient privileges to modify the server configuration
3. maybe there is a line of code I can change in some script, but I have no idea where

so I have no idea how to fix this.

**Comment by BeckeBauer** [comment]
Have you tried
`/usr/bin/php7.3 composer install --no-dev --optimize-autoloader`?

**Reply by RayCulp** [reply]
No, I have to admit I haven't tried using the full path up till now because simply using `php7.3` with other scripts was sufficient. But I'll be glad to try it. In any case, I've restored a backup now, so I can make a fresh attempt.

**Reply by RayCulp** [reply]
I just tried it and it gives the error: `"Could not open input file: composer"` Now that I see that error message, I think I actually did try that at some point. Maybe if I can locate the composer script and provide the full path to that as well. I try that.

**Reply by kevinpapst** [reply]
The docs I linked mention commands things like these:
```
php7.3.5-cli composer.phar install --no-dev --optimize-autoloader
php74 /usr/bin/composer install --no-dev --optimize-autoloader
/usr/bin/php7.3-cli composer.phar install --no-dev --optimize-autoloader
```

With `which composer` you get the full path. So try 
```
php7.3 `which composer` install --no-dev --optimize-autoloader
```

**Reply by RayCulp** [reply]
I understand now. The missing piece of the puzzle was that Composer is indeed a PHP script, and that by calling `composer install --no-dev --optimize-autoloader` I am actually implicitly invoking the PHP interpreter before that command. It's not an executable program or something else. I also just learned that a *.phar file is a PHP archive. I got it working with `/usr/bin/php7.3 /var/www/[accountname]/html/composer/composer.phar install --no-dev --optimize-autoloader`. Thank you!
